tribits/ci_support/CDashAnalyzeReportRandomFailures.py

```diff
@@ -0,0 +1,339 @@
+      dateUrlField+"&"+initialNonpassingTestFilters
+
+    # A.2) Create starting email body and html string aggregation var
+
+    cdashReportData = CDQAR.CDashReportData()
+
+    cdashReportData.htmlEmailBodyTop += \
+      "<h2>Random test failure scan results for "+cdashProjectName\
+        +" from "+dateRangeStr+"</h2>\n\n"
+
+
+    # B.1) Get all failing test result for past daysOfHistory
+
+    # Beginning of scanned details and links paragraph
+    cdashReportData.htmlEmailBodyTop +="<p>\n"
+
+    print("\nGetting list of initial nonpassing tests from CDash from "+dateRangeStr)
+
+    initialNonpassingTestsQueryUrl = CDQAR.getCDashQueryTestsQueryUrl(
+      cdashSiteUrl, cdashProjectName, None, initialNonpassingTestQueryFilters)
+    initialNonpassingTestBrowserUrl = CDQAR.getCDashQueryTestsBrowserUrl(
+      cdashSiteUrl, cdashProjectName, None, initialNonpassingTestQueryFilters)
+
+    if printUrlMode == 'initial' or printUrlMode == 'all':
+      print("\nCDash nonpassing tests browser URL:\n\n"+\
+        "  "+initialNonpassingTestBrowserUrl+"\n")
+      print("\nCDash nonpassing tests query URL:\n\n"+\
+        "  "+initialNonpassingTestsQueryUrl+"\n")
+
+    initialNonpassingTestsQueryCacheFile = testQueriesCacheDir +\
+      "/initialCDashNonPassingTests_"+dateRangeStart+"_"+dateRangeEnd+".json"
+
+    # List of dictionaries containing the cdash results in rows
+    initialNonpassingTestsLOD = CDQAR.downloadTestsOffCDashQueryTestsAndFlatten(
+      initialNonpassingTestsQueryUrl, initialNonpassingTestsQueryCacheFile,\
+      alwaysUseCacheFileIfExists=True)
+
+    cdashReportData.htmlEmailBodyTop += \
+      "<a href=\""+initialNonpassingTestBrowserUrl+"\">" +\
+      "Nonpassing tests scanned on CDash</a>=" +\
+      str(len(initialNonpassingTestsLOD))+"<br>\n"
+
+    # Ending of scanned details and links paragraph
+    # and start of scanning summaries and table
+    cdashReportData.htmlEmailBodyTop +="</p>\n\n<p>\n"
+
+    # B.2) Get each nonpassing test's testing history
+    for nonpassingTest in initialNonpassingTestsLOD:
+
+      # Remove unique jenkins run ID from build name
+      correctedBuildName = self.extractBuildNameStrategy.getCoreBuildName(nonpassingTest['buildName'])
+
+      testNameBuildName = nonpassingTest['testname']+"_"+nonpassingTest['buildName']
+      testNameBuildName = CDQAR.getCompressedFileNameIfTooLong(testNameBuildName)
+
+      print("\n Getting history from "+dateRangeStr+" for\n"+\
+            "  Test name: "+nonpassingTest['testname']+"\n"+\
+            "  Build name: "+correctedBuildName)
+
+      groupNameNormUrl, = CDQAR.normalizeUrlStrings(groupName)
+
+      testHistoryFilters = \
+        "filtercount=3&showfilters=1&filtercombine=and"+\
+        "&field1=testname&compare1=63&value1="+nonpassingTest['testname']+\
+        "&field2=groupname&compare2=63&value2="+groupNameNormUrl+\
+        "&field3=buildname&compare3=63&value3="+correctedBuildName
```

Comment:
In the future, it might be good to put this into a small free function given the input data?